File: c/src/ui/panel.c
#include "panel.h"
#include <stdio.h>

void panel_add_item(Panel *p, PanelItem item) {
    if (p == NULL) return;
    list_append(&p->items, item);
}

void panel_empty_line(Panel *p) {
    if (p == NULL) return;
    PanelItem item = {.ctx = NULL, .draw = NULL};
    list_append(&p->items, item);
}

void panel_draw(Panel *p, Screen *s, BBox b) {
    if (p == NULL || s == NULL) return;

    for (int i = 0; i < p->items.count && (u32)i < b.h; i++) {
        if (p->items.items[i].draw) {
            p->items.items[i].draw(p->items.items[i].ctx, s, b.x, b.y + i);
        }
    }
}

void panel_free(Panel *p) {
    list_free(p->items);
}

static void panel_draw_frame(void *ctx, Screen *s, BBox bbox) {
    Panel *p = (Panel *)ctx;
    panel_draw(p, s, bbox);
}

Widget panel_as_widget(Panel *p) {
    Widget widget = {
        .ctx = (void *)p,
        .draw = panel_draw_frame,
        .free_item = (WidgetFreeFn)panel_free,
    };
    return widget;
}

void panel_item_text_draw(void *ctx, Screen *s, int x, int y) {
    char *text = (char *)ctx;
    screen_puts(s, x, y, text);
}

void value_draw(void *ctx, Screen *s, int x, int y) {
    Value *v = (Value *)ctx;
    static char label_buf[256];
    static char value_buf[256];

    // Get stringified value
    if (v->stringify) {
        v->stringify(v->ctx, value_buf, sizeof(value_buf));
    } else {
        value_buf[0] = '\0';
    }

    snprintf(label_buf, sizeof(label_buf), "%s: %s", v->label, value_buf);
    screen_puts(s, x, y, label_buf);
}

PanelItem value_as_panel_item(Value *v) {
    PanelItem item = {
        .ctx = (void *)v,
        .draw = value_draw,
    };
    return item;
}

void value_stringify_int(void *ctx, char *buf, usize buf_size) {
    int *val = (int *)ctx;
    snprintf(buf, buf_size, "%d", *val);
}

void value_stringify_percent(void *ctx, char *buf, usize buf_size) {
    int *val = (int *)ctx;
    snprintf(buf, buf_size, "%d%%", *val);
}

void value_stringify_time(void *ctx, char *buf, usize buf_size) {
    int *val = (int *)ctx;
    int secs = *val;
    int mins = secs / 60;
    secs = secs % 60;
    if (mins > 0) {
        snprintf(buf, buf_size, "%d:%02d", mins, secs);
    } else {
        snprintf(buf, buf_size, "%ds", secs);
    }
}

File: rust/src/ui/panel.rs
//! Stats panel with dynamic values using closures

use crate::tui::screen::Screen;
use crate::ui::widget::{BBox, Widget};

/// Panel item types
pub enum PanelItem {
    Empty,
    Text(String),
    Value {
        label: String,
        get_value: Box<dyn Fn() -> String + Send + Sync>,
    },
}

/// Stats panel widget
pub struct Panel {
    items: Vec<PanelItem>,
}

#[allow(unused)]
impl Panel {
    pub fn new() -> Self {
        Panel { items: Vec::new() }
    }

    pub fn add_empty_line(&mut self) {
        self.items.push(PanelItem::Empty);
    }

    pub fn add_text(&mut self, text: &str) {
        self.items.push(PanelItem::Text(text.to_string()));
    }

    pub fn add_value<F>(&mut self, label: &str, get_value: F)
    where
        F: Fn() -> String + Send + Sync + 'static,
    {
        self.items.push(PanelItem::Value {
            label: label.to_string(),
            get_value: Box::new(get_value),
        });
    }
}

impl Default for Panel {
    fn default() -> Self {
        Self::new()
    }
}

impl Widget for Panel {
    fn draw(&self, screen: &mut Screen, bbox: BBox) {
        for (i, item) in self.items.iter().enumerate() {
            if i as u32 >= bbox.h {
                break;
            }

            let x = bbox.x as usize;
            let y = (bbox.y + i as u32) as usize;

            match item {
                PanelItem::Empty => {}
                PanelItem::Text(text) => {
                    screen.put_str(x, y, text);
                }
                PanelItem::Value { label, get_value } => {
                    let value = get_value();
                    let text = format!("{}: {}", label, value);
                    screen.put_str(x, y, &text);
                }
            }
        }
    }
}
